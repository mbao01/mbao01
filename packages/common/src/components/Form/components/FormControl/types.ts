import { type VariantProps } from "../../../../libs";
import { getFormControlClasses } from "./constants";

export type As = "div" | "span" | "label";

type AllowedElements = Pick<React.JSX.IntrinsicElements, As>;

export type FormControlProps<T extends As> = (T extends "label"
  ? React.LabelHTMLAttributes<HTMLLabelElement>
  : React.HTMLAttributes<HTMLElement & AllowedElements[T]>) &
  VariantProps<typeof getFormControlClasses> & {
    as?: T;
  };
