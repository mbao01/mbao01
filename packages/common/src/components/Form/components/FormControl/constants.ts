import { cva } from "../../../../libs";

export const getFormControlClasses = cva("form-control w-full text-sm");
