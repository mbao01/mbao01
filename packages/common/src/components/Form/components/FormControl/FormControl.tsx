import { Slot } from "@radix-ui/react-slot";
import type { As, FormControlProps } from "./types";
import { getFormControlClasses } from "./constants";
import { cn } from "../../../../utilities";

export const FormControl = <T extends As>({
  as,
  children,
  className,
  ...props
}: FormControlProps<T>) => {
  const SlotChild = as ?? "label";

  return (
    <Slot className={cn(getFormControlClasses(), className)} {...props}>
      {/* slot merges it's prop into it's immediate child */}
      <SlotChild>{children}</SlotChild>
    </Slot>
  );
};
