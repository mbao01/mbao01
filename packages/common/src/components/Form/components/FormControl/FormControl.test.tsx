import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { FormControl } from "./FormControl";

describe("FormControl", () => {
  it("can be a label form control element", () => {
    const { asFragment } = render(
      <FormControl as="label" htmlFor="username">
        Username
        <input id="username" name="username" />
      </FormControl>
    );

    expect(screen.getByLabelText("Username")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it.each(["div", "span"] as const)("can be a %s element", (as) => {
    const description = `${as} element`;
    const { asFragment } = render(
      <FormControl as={as}>{description}</FormControl>
    );

    expect(screen.getByText(description)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
