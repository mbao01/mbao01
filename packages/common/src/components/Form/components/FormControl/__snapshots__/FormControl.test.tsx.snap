// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormControl > can be a div element 1`] = `
<DocumentFragment>
  <div
    class="form-control w-full"
  >
    div element
  </div>
</DocumentFragment>
`;

exports[`FormControl > can be a label form control element 1`] = `
<DocumentFragment>
  <label
    class="label form-control w-full"
    for="username"
  >
    Username
    <input
      id="username"
      name="username"
    />
  </label>
</DocumentFragment>
`;

exports[`FormControl > can be a span element 1`] = `
<DocumentFragment>
  <span
    class="form-control w-full"
  >
    span element
  </span>
</DocumentFragment>
`;
